subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java

```diff
@@ -79,19 +90,50 @@ public EclipseDependenciesCreator(EclipseClasspath classpath) {
         boolean downloadSources = classpath.isDownloadSources();
         boolean downloadJavadoc = classpath.isDownloadJavadoc();
 
+        Map<String, Set<String>> pathToSourceSets = collectRuntimeClasspathPathSourceSets();
+
         Collection<IdeExtendedRepoFileDependency> repoFileDependencies = dependenciesExtractor.extractRepoFileDependencies(classpath.getProject().getDependencies(), classpath.getPlusConfigurations(), classpath.getMinusConfigurations(), downloadSources, downloadJavadoc);
         for (IdeExtendedRepoFileDependency dependency : repoFileDependencies) {
-            libraries.add(createLibraryEntry(dependency.getFile(), dependency.getSourceFile(), dependency.getJavadocFile(), classpath, dependency.getId()));
+            libraries.add(createLibraryEntry(dependency.getFile(), dependency.getSourceFile(), dependency.getJavadocFile(), classpath, dependency.getId(), pathToSourceSets));
         }
 
         Collection<IdeLocalFileDependency> localFileDependencies = dependenciesExtractor.extractLocalFileDependencies(classpath.getPlusConfigurations(), classpath.getMinusConfigurations());
         for (IdeLocalFileDependency it : localFileDependencies) {
-            libraries.add(createLibraryEntry(it.getFile(), null, null, classpath, null));
+            libraries.add(createLibraryEntry(it.getFile(), null, null, classpath, null, pathToSourceSets));
         }
         return libraries;
     }
 
-    private static AbstractLibrary createLibraryEntry(File binary, File source, File javadoc, EclipseClasspath classpath, ModuleVersionIdentifier id) {
+    private Map<String, Set<String>> collectRuntimeClasspathPathSourceSets() {
+        Map<String, Set<String>> pathToSourceSetNames = Maps.newHashMap();
+        Iterable<SourceSet> sourceSets = classpath.getSourceSets();
+
+        // for non-java projects there are no source sets configured
+        if (sourceSets == null) {
+            return pathToSourceSetNames;
+        }
+
+        try {
+            for (SourceSet sourceSet : sourceSets) {
+                String name = sourceSet.getName().replace(",", "");
+                FileCollection classpath = sourceSet.getRuntimeClasspath();
```

Comment:
❌ We need to check both compile and runtime classpath.

Reply:
Fixed.